Initial interfaces

XedinUnknown opened the pull request.

XedinUnknown committed:
Now depending on `dhii/validation-interface`

XedinUnknown committed:
Added `TypeInterface`

XedinUnknown committed:
Added `ScalableTypeInterface` and test

XedinUnknown committed:
Added `CompountTypeInterface`

XedinUnknown committed:
Added `PseudoTypeInterface` and test

mecha commented on src/CompoundTypeInterface.php:
```diff
@@ -0,0 +1,25 @@
+ *
+ * @since [*next-version*]
+ */
+interface CompoundTypeInterface extends TypeInterface
```
Consider different naming, such as `StructureTypeInterface`, since "compound" means made up of two or more elements, which is not a restriction for these types.

XedinUnknown replied:
@mecha, taken from [PHP manual](http://php.net/manual/en/language.types.intro.php).

mecha replied:
Interesting. Well then stick to PHP's terminology for the sake of consistency 👍

mecha reviewed (commented):
I'm not a big fan of the 3-letter naming scheme, especially since the primitive type constants are longer (`BOOL`, `FLOAT`, ...). Is there a particular reason for this? Perhaps in anticipation of some parsing strategy?

XedinUnknown commented:
@mecha, I don't observe a particularly 3-letter naming scheme in the package. Some abbreviations are 3-letter long, because that's a very usual way to name those things. However, there isn't any other particular reason, and no other names follows that. Would you like to suggest different names?

mecha commented:
I think `SCAL`, `ITER`, `NUM` and `RES` should be full words, unless their full word versions are PHP keywords. I understand `NUL` and `CALL`, as those become PHP keywords.

Alternatively, consider prefixing the constants with `T_`

I think the consistency and intuitive names will be of great benefit; i.e. I don't have to remember what constant name a `callable` has. I just have to use `T_CALLABLE`.

That's just my 2 cents

XedinUnknown committed:
Changed type constant naming convention

All type constant names are now prefixed with `T_`, and have the
full name of the type, as opposed to an abbreviation or acronym.

XedinUnknown committed:
Tests updated to reflect naming convention change

mecha reviewed (approved):
So nice 👍

mecha commented on src/TypeInterface.php:
```diff
@@ -0,0 +1,24 @@
+     *
+     * @since [*next-version*]
+     *
+     * @param bool $strict Whether or not comparison will be done in strict mode.
```
Missing param doc for `$value`, or missing `{@inheritdoc}`

XedinUnknown commented:
During a voice call, it was decided that a `KeyAwareTypeInterface` can be useful, and must be included in this PR.

XedinUnknown committed:
Now depending also on key-value aware interface

All deps updated

XedinUnknown committed:
Added `KeyAwareTypeInterface` and test

mecha commented on test/unit/KeyAwareTypeInterfaceTest.php:
```diff
@@ -0,0 +1,50 @@
+
+        $this->assertInstanceOf(static::TEST_SUBJECT_CLASSNAME, $subject, 'A valid instance of the test subject could not be created');
+        $this->assertInstanceOf('Dhii\Data\KeyAwareInterface', $subject, 'Subject does not implement required interface');
+    }
```
Also assert whether the subject is an instance of `TypeInterface`

mecha commented:
@XedinUnknown Also consider including a `TypeAwareInterface`.

XedinUnknown committed:
`KeyAwareTypeInterface` test now ascertains that it's a `TypeInterface`

XedinUnknown committed:
Added `TypeAwareInterface`

XedinUnknown committed:
Added test for `TypeAwareInterface`

mecha commented on src/PseudoTypeInterface.php:
```diff
@@ -0,0 +1,85 @@
+     *
+     * @since [*next-version*]
+     */
+    const T_SCALABLE = 'scalable';
```
Isn't this supposed to be "scalar" not "scalable"?

mecha commented on src/ScalableTypeInterface.php:
```diff
@@ -0,0 +1,47 @@
+ *
+ * @since [*next-version*]
+ */
+interface ScalableTypeInterface extends TypeInterface
```
Same here. I think the proper name is "scalar" not "scalable".

Unfortunately due to the use of the word "scalable" in terms of web servers and PHP being a web-focused language, I can't find anything related to types.

But [on the manual they use "scalar"](http://php.net/manual/en/language.types.intro.php).

XedinUnknown committed:
Dependencies updated

XedinUnknown committed:
Renamed `Scalable` to `Scalar`

Also fixed some tests and docs

XedinUnknown committed:
Aliased `develop` to `0.1.x-dev`

XedinUnknown committed:
Changed "scalable" to "scalar" in pseudo type and tests

XedinUnknown committed:
Added missing param doc

mecha reviewed (approved):
LGTM

XedinUnknown committed:
Now building for PHP 5.3 also
